UefiTestingPkg/AuditTests/PagingAudit/UEFI/Dxe/App/ProjectMuFunctionality.c

```diff
@@ -0,0 +1,332 @@
+#include "DxePagingAuditTestApp.h"
+#include <Protocol/MemoryProtectionSpecialRegionProtocol.h>
+#include <Protocol/MemoryProtectionDebug.h>
+#include <Protocol/CpuMpDebug.h>
+#include <Protocol/MemoryProtectionDebug.h>
+#include <Library/DxeMemoryProtectionHobLib.h>
+#include <Library/HobLib.h>
+
+MEMORY_PROTECTION_SPECIAL_REGION  *mSpecialRegions           = NULL;
+IMAGE_RANGE_DESCRIPTOR            *mNonProtectedImageList    = NULL;
+UINTN                             mSpecialRegionCount        = 0;
+MEMORY_PROTECTION_DEBUG_PROTOCOL  *mMemoryProtectionProtocol = NULL;
+CPU_MP_DEBUG_PROTOCOL             *mCpuMpDebugProtocol       = NULL;
+
+/**
+  Populates the heap guard protocol global
+
+  @retval EFI_SUCCESS Protocol is already populated or was successfully populated
+  @retval other       Return value of LocateProtocol
+**/
+EFI_STATUS
+PopulateHeapGuardDebugProtocol (
+  VOID
+  )
+{
+  if (mMemoryProtectionProtocol != NULL) {
+    return EFI_SUCCESS;
+  }
+
+  return gBS->LocateProtocol (&gMemoryProtectionDebugProtocolGuid, NULL, (VOID **)&mMemoryProtectionProtocol);
+}
+
+/**
+  Populates the CPU MP debug protocol global
+
+  @retval EFI_SUCCESS Protocol is already populated or was successfully populated
+  @retval other       Return value of LocateProtocol
+**/
+EFI_STATUS
+PopulateCpuMpDebugProtocol (
+  VOID
+  )
+{
+  if (mCpuMpDebugProtocol != NULL) {
+    return EFI_SUCCESS;
+  }
+
+  return gBS->LocateProtocol (&gCpuMpDebugProtocolGuid, NULL, (VOID **)&mCpuMpDebugProtocol);
+}
+
+/*
+  Writes the NULL page and stack information to the memory info database
+ */
+VOID
+ProjectMuSpecialMemoryDump (

```

Comment:
Same about not associating implementation functionality with a firmware project.